### PR TITLE
Return immediately from wait if at least one invalid channel

### LINKED ISSUE #803: `wait_on_channels` ABI call should fail immediately if any channel handle is invalid
Currently it just ignores invalid handles, unless all of them are invalid, in which case it does return an error.

See https://github.com/project-oak/oak/blob/c2b2ff4b666de2a4f5fe7ff7cbfadbfc00af69f2/docs/abi.md#wait_on_channels

This should also allow to simplify some of the Runtime logic dealing with optional handles for this ABI call:

https://github.com/project-oak/oak/blob/c2b2ff4b666de2a4f5fe7ff7cbfadbfc00af69f2/oak/server/rust/oak_runtime/src/runtime/mod.rs#L427-L431

### PR DESCRIPTION
**Pull request by rbehjati**
* `wait_on_channel` in `oak_runtime` returns early if any of the channels is invalid or orphaned
* `wait_on_channel` in `oak_runtime` returns a vector containing all statuses if any is inaccessible
* Added the a new status `PERMISSION_DENIED` to `ChannelReadStatus` 
* Updated documentation

Fixes #803 

# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [x] I have updated [documentation](docs/) accordingly.
  - [x] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

**Comment by rbehjati** [comment]
@tiziano88  @daviddrysdale What do you think about this one? After David merged his changes (that removed the indirection), this has become simpler. 
I also decided not to change the signature, and so did not have to change anything in the `SDK` nor in the `wasm` node. It is just that the runtime returns earlier, but still with all the statuses.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 1]
```diff
@@ -799,15 +782,18 @@ impl Runtime {
...
     /// - If the [`Runtime`] is terminating this will return immediately with an `ErrTerminated`
     ///   status for each channel.
-    /// - If all readers are in an erroneous status, e.g. when all channels are orphaned, this will
-    ///   immediately return the channels statuses.
+    /// - If any of the readers are in an erroneous status, e.g. when a channel is orphaned, this
+    ///   will immediately return with all the channels statuses set in the returned vector.
     /// - If any of the channels is able to read a message, the corresponding element in the
     ///   returned vector will be set to `Ok(ReadReady)`, with `Ok(NotReady)` signaling the channel
-    ///   has no message available
+    ///   has no message available.
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:789]: There is no overall value. In both cases we return a vector of all updated statuses immediately after the status change happens. The caller code decides how to handle the situation. (I am assuming any status change will unpark the thread, thus re-executing the loop and updating all the statuses)

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 2]
```diff
@@ -826,10 +812,31 @@ impl Runtime {
             }
         }
 
-        self.validate_can_read_from_channels(node_id, &readers)?;
+        if readers.is_empty() {
```
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs]: Good catch. I have fixed it, and updated the test too.

File: docs/abi.md [hunk 3]
```diff
@@ -87,17 +87,19 @@ functions** as
...
-from one of the specified channel handles. The channel handles are encoded in a
-buffer that holds N contiguous 9-byte chunks, each of which is made up of an
-8-byte channel handle value (little-endian u64) followed by a single channel
-status byte. Invalid handles will have an `INVALID_CHANNEL` status, but
+from one of the specified channel handles, unless any of the channels is
+invalid, orphaned, or violates the
+[information flow control](/docs/concepts.md#labels). The channel handles are
+encoded in a buffer that holds N contiguous 9-byte chunks, each of which is made
+up of an 8-byte channel handle value (little-endian u64) followed by a single
+channel status byte. Invalid and orphaned handles will have an `INVALID_CHANNEL`
```
Comment (tiziano88) [inline docs/abi.md:95]: I would still make it very clear that the overall return value in this case is still `OK`, so callers need to be careful about checking the individual channel statuses (which I guess is fair enough because in general they would want to know which channel - if any - actually got ready).

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 4]
```diff
@@ -849,18 +834,21 @@ impl Runtime {
                 })?;
             }
             let statuses = self.readers_statuses(node_id, &readers);
+            // Transcribe the status for valid channels back to the original position
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:837]: I only had a quick look, but it seems that the logic above is still just ignoring all initially invalid channels?

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 5]
```diff
@@ -849,18 +834,21 @@ impl Runtime {
...
+            // in the list of all statuses.
+            for i in 0..readers.len() {
+                all_statuses[reader_pos[i]] = statuses[i];
+            }
 
-            let all_unreadable = statuses.iter().all(|&s| {
-                s == ChannelReadStatus::InvalidChannel || s == ChannelReadStatus::Orphaned
+            let any_invalid = statuses.iter().any(|&s| {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:843]: Given the newly updated documentation, I would expect simply that if any reader status is different from NotReady, then the entire vector is returned immediately. Or did I misunderstand?

File: docs/abi.md (hunk 3 above)
Comment (rbehjati) [reply docs/abi.md]: The sentence after this states the condition for `error`, which I think is clear enough. Nevertheless, I have reworded this sentence to look more like the original version.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs (hunk 5 above)
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs]: Yes. That is the new logic. So, it waits only if all statuses are `NotReady`. I could change the condition to check that instead. It would be clearer in my opinion.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs (hunk 4 above)
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:837]: I think `readers.len() != read_handles.len()` in line `851` checks this condition and returns immediately. It only waits incorrectly if the input vector is empty! I have fixed it now.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 6]
```diff
@@ -799,15 +782,19 @@ impl Runtime {
     /// Waits on a slice of `ChannelHalf`s, blocking until one of the following conditions:
     /// - If the [`Runtime`] is terminating this will return immediately with an `ErrTerminated`
     ///   status for each channel.
-    /// - If all readers are in an erroneous status, e.g. when all channels are orphaned, this will
-    ///   immediately return the channels statuses.
+    /// - If any of the readers is in an erroneous status, e.g. when a channel is orphaned, this
+    ///   will immediately return with all the channels statuses set in the returned vector.
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:786]: (pre-existing) Maybe reword the entire comment, since it kind of contradicts itself saying that the method blocks until one of these conditions. Make it clear under what circumstances it returns immediately and when it blocks, and when blocked, what condition(s) unblock it, and what is the return value for each case.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs (hunk 4 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:837]: Ah yes you are right, though the logic seems unnecessarily convoluted now. Perhaps something to try and simplify in a separate PR?

File: oak/server/rust/oak_runtime/src/runtime/tests.rs [hunk 7]
```diff
@@ -140,3 +140,68 @@ fn create_node_more_public_label() {
...
     );
 }
+
+#[test]
+fn wait_on_channels_immediately_returns_if_any_channel_is_orphaned() {
+    run_node_body(
+        Label::public_trusted(),
+        Box::new(|runtime| {
+            // Create channels with handles
+            let (write_handle, read_handle_1) = runtime.channel_create(&Label::public_trusted());
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/tests.rs:150]:
```suggestion
            let (write_handle_1, read_handle_1) = runtime.channel_create(&Label::public_trusted());
```

for consistency (and `_2` below).

Also in other places we start from 0, but up to you really.

File: oak/server/rust/oak_runtime/src/runtime/tests.rs [hunk 8]
```diff
@@ -140,3 +140,68 @@ fn create_node_more_public_label() {
...
+        }),
+    );
+}
+
+#[test]
+fn wait_on_channels_immediately_returns_if_the_input_list_is_empty() {
+    run_node_body(
+        Label::public_trusted(),
+        Box::new(|runtime| {
+            // Wait on channel
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/tests.rs:201]: Seems unnecessary, also there is no channel here, right?

File: oak/server/rust/oak_runtime/src/runtime/tests.rs [hunk 9]
```diff
@@ -140,3 +140,68 @@ fn create_node_more_public_label() {
...
+        Box::new(|runtime| {
+            // Create channels with handles
+            let (write_handle, read_handle_1) = runtime.channel_create(&Label::public_trusted());
+            let (_write_handle, read_handle_2) = runtime.channel_create(&Label::public_trusted());
+
+            // Close the write_handle; this should make the channel Orphaned
+            let result = runtime.clone().channel_close(write_handle);
+            assert_eq!(Ok(()), result);
+
+            // Wait on channel
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/tests.rs:157]: Here there are two channels. Also I don't think this comment is particularly relevant.

File: oak/server/rust/oak_runtime/src/runtime/tests.rs [hunk 10]
```diff
@@ -140,3 +140,68 @@ fn create_node_more_public_label() {
...
+            let result = runtime
+                .clone()
+                .wait_on_channels(&[read_handle_1, read_handle_2]);
+            assert!(result.is_ok());
+            assert_eq!(
+                vec![ChannelReadStatus::Orphaned, ChannelReadStatus::NotReady],
+                result?
+            );
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/tests.rs:165]: (optional) Does it work to just match the entire thing at once?

```rust
assert_eq!(Ok(vec![ChannelReadStatus::Orphaned, ChannelReadStatus::NotReady]), result)
```

also elsewhere

File: oak/server/rust/oak_runtime/src/runtime/tests.rs [hunk 11]
```diff
@@ -140,3 +140,68 @@ fn create_node_more_public_label() {
         }),
     );
 }
+
+#[test]
+fn wait_on_channels_immediately_returns_if_any_channel_is_orphaned() {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/tests.rs:145]: Since the interesting part about the wait functionality is the waiting, would it be possible to have a test that does something concurrently? e.g. create two nodes, and close a channel after a while to make it orphaned, or send something on one of the channels while another node is waiting? I am not sure whether the test harness we currently have supports this, we may have to extend it to support running multiple nodes (maybe two nodes is enough?). Something for another PR perhaps, but could you create an issue and look into it @rbehjati ?

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 12]
```diff
@@ -953,7 +935,11 @@ impl Runtime {
         node_id: NodeId,
         half: &ChannelHalf,
     ) -> Result<ChannelReadStatus, OakStatus> {
-        self.validate_can_read_from_channel(node_id, half)?;
+        if let Err(OakStatus::ErrPermissionDenied) =
+            self.validate_can_read_from_channel(node_id, half)
+        {
+            return Ok(ChannelReadStatus::PermissionDenied);
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:941]: Update method documentation.

Also could you clarify when an overall `Err` is returned (if at all)?

File: oak/server/rust/oak_runtime/src/runtime/tests.rs (hunk 11 above)
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/tests.rs:135]: I have added a new test that closes the write end of the channel in a separate thread. WDYT?

File: oak/server/rust/oak_runtime/src/runtime/tests.rs (hunk 9 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/runtime/tests.rs]: Oh I just meant that maybe it should be channels (plural). But also that it also seems somewhat obvious (usually comments are for "why", not "what"), but feel free to keep it if you think it helps explain the flow.

File: oak/server/rust/oak_runtime/src/runtime/tests.rs (hunk 11 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/runtime/tests.rs:135]: Very neat! Thanks for doing this! Maybe we should think about a framework to make this kind of test even simpler in the future, e.g. running multiple node bodies in parallel. If you have any ideas, please create some GitHub issues to track.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 13]
```diff
@@ -955,15 +940,23 @@ impl Runtime {
     }
 
     /// Determine the readable status of a channel, returning:
-    /// - [`ChannelReadStatus::ReadReady`] if there is at least one message in the channel.
-    /// - [`ChannelReadStatus::Orphaned`] if there are no messages and there are no writers
-    /// - [`ChannelReadStatus::NotReady`] if there are no messages but there are some writers
+    /// - [`Ok(ChannelReadStatus::ReadReady)`] if there is at least one message in the channel.
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:943]: I think these are not valid links now, are they?

Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs]: You are right. Thanks. I have fixed them now. Although this function is not public, so at the moment, the documentation is not being generated for it.

File: oak/server/rust/oak_runtime/src/runtime/tests.rs (hunk 11 above)
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/tests.rs:135]: I agree. It would be more realistic to simulate the situation in this test with two parallel nodes. I'll think about it, especially if there are other scenarios that could benefit from such a utility.

File: docs/abi.md [hunk 14]
```diff
@@ -89,16 +89,14 @@ functions** as
...
-[information flow control](/docs/concepts.md#labels), returns
-`ERR_PERMISSION_DENIED`.
+from one of the specified channel handles, unless any of the channels is
+invalid, orphaned, or violates the
+[information flow control](/docs/concepts.md#labels). The channel handles are
+encoded in a buffer that holds N contiguous 9-byte chunks, each of which is made
+up of an 8-byte channel handle value (little-endian u64) followed by a single
+channel status byte. Invalid handles will have an `INVALID_CHANNEL`, `ORPHANED`,
+or `PERMISSION_DENIED` status, but `wait_on_channels` return value will only
+fail for internal errors or if _all_ channels are invalid.
```
Comment (daviddrysdale) [inline docs/abi.md:99]: Is this text still correct?

File: oak/server/rust/oak_runtime/src/runtime/tests.rs (hunk 11 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/runtime/tests.rs:135]: In "userspace" there's already [functionality for testing multi-Node applications](
https://github.com/project-oak/oak/blob/master/docs/programming-oak.md#testing-multi-node-applications); probably some commonization/re-use is possible.

File: docs/abi.md (hunk 14 above)
Comment (rbehjati) [reply docs/abi.md:99]: I think so, since this is about the `ABI`. But it is not the case for `wait_on_channels` in the `Runtime`. There is an extra check in the implementation of `wait_on_channels` in the `wasm` node that returns `Err(OakStatus::ErrBadHandle)` if all channels are invalid (perhaps I could update it to check for `ChannelReadStatus::PermissionDenied` too). 
```
        if statuses
            .iter()
            .all(|&s| s == ChannelReadStatus::InvalidChannel || s == ChannelReadStatus::Orphaned)
        {
            Err(OakStatus::ErrBadHandle)
        } else {
            Ok(())
        }
```

In this case the return value is just an extra flag. We already return all the updated statuses in the `status_buff` (arg 0).

File: oak/server/rust/oak_runtime/src/runtime/tests.rs (hunk 11 above)
Comment (rbehjati) [reply oak/server/rust/oak_runtime/src/runtime/tests.rs:135]: Nice. Thanks.

File: docs/abi.md (hunk 14 above)
Comment (daviddrysdale) [reply docs/abi.md:99]: Ah yes, I misread it &ndash; it's describing the return value, not the conditions under which the call blocks (which is at the start).
